File: 03321. Find X-Sum of All K-Long Subarrays II/solution.go
/**
 * Problem: 3321. Find X-Sum of All K-Long Subarrays II
 *
 * Difficulty: Hard
 *
 * Language: Golang
 *
 * Performance: Runtime - 214 ms (Beats 100%)
 */

// Define a struct `item` to represent a value, its frequency, index in the heap,
// and whether it is part of the top `x` elements.
type item struct {
	val   int
	freq  int
	idx   int
	inTop bool
}

// Define a min-heap `hotHeap` for the top `x` elements.
type hotHeap []*item

// Return the length of the `hotHeap`.
func (h hotHeap) Len() int { return len(h) }

// Compare two items in the `hotHeap` based on frequency and value.
func (h hotHeap) Less(i, j int) bool {
	if h[i].freq != h[j].freq {
		return h[i].freq < h[j].freq
	}
	return h[i].val < h[j].val
}

// Swap two items in the `hotHeap` and update their indices.
func (h hotHeap) Swap(i, j int) { h[i], h[j] = h[j], h[i]; h[i].idx, h[j].idx = i, j }

// Push a new item into the `hotHeap` and set its index.
func (h *hotHeap) Push(x interface{}) {
	it := x.(*item)
	it.idx = len(*h)
	*h = append(*h, it)
}

// Pop the top item from the `hotHeap` and return it.
func (h *hotHeap) Pop() interface{} {
	old := *h
	it := old[len(old)-1]
	*h = old[:len(old)-1]
	return it
}

// Define a max-heap `restHeap` for the remaining elements.
type restHeap []*item

// Return the length of the `restHeap`.
func (h restHeap) Len() int { return len(h) }

// Compare two items in the `restHeap` based on frequency and value.
func (h restHeap) Less(i, j int) bool {
	if h[i].freq != h[j].freq {
		return h[i].freq > h[j].freq
	}
	return h[i].val > h[j].val
}

// Swap two items in the `restHeap` and update their indices.
func (h restHeap) Swap(i, j int) { h[i], h[j] = h[j], h[i]; h[i].idx, h[j].idx = i, j }

// Push a new item into the `restHeap` and set its index.
func (h *restHeap) Push(x interface{}) {
	it := x.(*item)
	it.idx = len(*h)
	*h = append(*h, it)
}

// Pop the top item from the `restHeap` and return it.
func (h *restHeap) Pop() interface{} {
	old := *h
	it := old[len(old)-1]
	*h = old[:len(old)-1]
	return it
}

// Function `findXSum` calculates the X-sum of all K-long subarrays.
func findXSum(nums []int, k int, x int) []int64 {
	// Initialize the length of the input array and result array.
	n := len(nums)
	ans := make([]int64, n-k+1)

	// Create a map to store frequency of values and initialize heaps.
	freq := map[int]*item{}
	hot := &hotHeap{}
	rest := &restHeap{}

	// Initialize the heaps.
	heap.Init(hot)
	heap.Init(rest)

	// Variable to store the sum of the top `x` elements.
	var sum int64 = 0

	// Function to remove an item from its current heap.
	removeFromCurrent := func(it *item) {
		if it.inTop {
			sum -= int64(it.val) * int64(it.freq)
			heap.Remove(hot, it.idx)
			it.inTop = false
		} else if it.freq > 0 && it.idx >= 0 && it.idx < rest.Len() && (*rest)[it.idx] == it {
			heap.Remove(rest, it.idx)
		}
	}

	// Function to promote items from `restHeap` to `hotHeap` if possible.
	promoteIfPossible := func() {
		for hot.Len() < x && rest.Len() > 0 {
			best := heap.Pop(rest).(*item)
			best.inTop = true
			sum += int64(best.val) * int64(best.freq)
			heap.Push(hot, best)
		}
	}

	// Function to insert a value into the heaps.
	insertVal := func(v int) {
		it, ok := freq[v]
		if !ok {
			it = &item{val: v, idx: -1}
			freq[v] = it
		} else {
			removeFromCurrent(it)
		}

		it.freq++
		it.inTop = true
		sum += int64(it.val) * int64(it.freq)
		heap.Push(hot, it)

		if hot.Len() > x {
			worst := heap.Pop(hot).(*item)
			sum -= int64(worst.val) * int64(worst.freq)
			worst.inTop = false
			heap.Push(rest, worst)
		}
	}

	// Function to erase a value from the heaps.
	eraseVal := func(v int) {
		it := freq[v]
		removeFromCurrent(it)

		it.freq--
		if it.freq == 0 {
			delete(freq, v)
			it.idx, it.inTop = -1, false
		} else {
			heap.Push(rest, it)
		}

		promoteIfPossible()
	}

	// Insert the first `k` elements into the heaps.
	for i := 0; i < k; i++ {
		insertVal(nums[i])
	}

	// Store the initial sum in the result array.
	ans[0] = sum

	// Slide the window over the array and update the result.
	for i := k; i < n; i++ {
		eraseVal(nums[i-k])
		insertVal(nums[i])
		ans[i-k+1] = sum
	}

	// Return the result array.
	return ans
}

File: 03321. Find X-Sum of All K-Long Subarrays II/solution.js
/**
 * Problem: 3321. Find X-Sum of All K-Long Subarrays II
 *
 * Difficulty: Hard
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 422 ms (Beats 100%)
 */

/**
 * Calculates the sum of the top `x` frequent elements in a sliding window of size `k`
 *
 * @param {number[]} nums - Input array
 * @param {number} k - Sliding window size
 * @param {number} x - Number of top frequent elements to sum
 *
 * @returns {number[]} - Sums for each sliding window
 */
const findXSum = (nums, k, x) => {
  const n = nums.length
  const res = []
  let sum = 0

  const freqMap = new Map(),
    top = new OrderedSet(),
    rest = new OrderedSet()

  for (let i = 0; i < n; i++) {
    const count = freqMap.get(nums[i]) || 0

    if (count > 0) {
      //
      if (rest.find(nums[i], count)) rest.delete(nums[i], count)
      //
      else {
        top.delete(nums[i], count)
        sum -= nums[i] * count
      }
    }

    freqMap.set(nums[i], count + 1)
    top.insert(nums[i], count + 1)
    sum += nums[i] * (count + 1)

    if (top.size > x) {
      const [minNum, minCount] = top.getMin()

      sum -= minNum * minCount
      rest.insert(minNum, minCount)
      top.delete(minNum, minCount)
    }

    if (i >= k) {
      const leftCount = freqMap.get(nums[i - k])

      //
      if (rest.find(nums[i - k], leftCount)) rest.delete(nums[i - k], leftCount)
      //
      else {
        top.delete(nums[i - k], leftCount)
        sum -= leftCount * nums[i - k]
      }

      freqMap.set(nums[i - k], leftCount - 1)

      if (leftCount - 1 > 0) rest.insert(nums[i - k], leftCount - 1)

      if (top.size < x && rest.size > 0) {
        const [maxNum, maxCount] = rest.getMax()

        sum += maxNum * maxCount
        top.insert(maxNum, maxCount)
        rest.delete(maxNum, maxCount)
      }
    }

    if (i >= k - 1) res.push(sum)
  }

  return res
}

/**
 *
 */
class RBTreeNode {
  /**
   *
   * @param key
   * @param value
   * @param nilNode
   */
  constructor(key, value, nilNode) {
    this.key = key
    this.value = value
    this.color = 'red'
    this.left = nilNode
    this.right = nilNode
    this.parent = nilNode
  }

  /**
   *
   */
  isRed() {
    return this.color === 'red'
  }
}

/**
 *
 */
class RBTree {
  constructor() {
    this.nil = new RBTreeNode(null, null, null)
    this.nil.color = 'black'
    this.root = this.nil
  }

  /**
   *
   * @param node1
   * @param node2
   */
  compare(node1, node2) {
    if (node1.value !== node2.value) return node1.value - node2.value

    return node1.key - node2.key
  }

  /**
   *
   * @param key
   * @param value
   */
  insert(key, value) {
    const z = new RBTreeNode(key, value, this.nil)

    let y = this.nil,
      x = this.root

    while (x !== this.nil) {
      y = x

      if (this.compare(z, x) < 0) x = x.left
      else x = x.right
    }

    z.parent = y

    if (y === this.nil) this.root = z
    else if (this.compare(z, y) < 0) y.left = z
    else y.right = z

    z.left = this.nil
    z.right = this.nil
    z.color = 'red'

    this.insertFixup(z)
  }

  /**
   *
   * @param key
   * @param value
   */
  delete(key, value) {
    let node = this.root
    let targetNode = null

    while (node !== this.nil) {
      const tempNode = new RBTreeNode(key, value, this.nil)

      if (this.compare(tempNode, node) === 0) {
        targetNode = node
        break
      }
      //
      else if (this.compare(tempNode, node) < 0) node = node.left
      else node = node.right
    }

    if (targetNode) this._deleteNode(targetNode)
  }

  /**
   *
   * @param node
   */
  _deleteNode(node) {
    let y = node
    let yOriginalColor = y.color
    let x

    if (node.left === this.nil) {
      x = node.right
      this.transplant(node, node.right)
    } else if (node.right === this.nil) {
      x = node.left
      this.transplant(node, node.left)
    } else {
      y = this.minimum(node.right)
      yOriginalColor = y.color
      x = y.right

      if (y.parent === node) x.parent = y
      //
      else {
        this.transplant(y, y.right)
        y.right = node.right
        y.right.parent = y
      }

      this.transplant(node, y)

      y.left = node.left
      y.left.parent = y
      y.color = node.color
    }

    if (yOriginalColor === 'black') this.deleteFixup(x)
  }

  /**
   *
   * @param u
   * @param v
   */
  transplant(u, v) {
    if (u.parent === this.nil) this.root = v
    else if (u === u.parent.left) u.parent.left = v
    else u.parent.right = v

    v.parent = u.parent
  }

  /**
   *
   * @param node
   */
  minimum(node) {
    while (node.left !== this.nil) node = node.left

    return node
  }

  /**
   *
   * @param node
   */
  maximum(node) {
    while (node.right !== this.nil) node = node.right

    return node
  }

  /**
   *
   * @param x
   */
  deleteFixup(x) {
    while (x !== this.root && !x.isRed()) {
      if (x === x.parent.left) {
        let w = x.parent.right

        if (w.isRed()) {
          w.color = 'black'
          x.parent.color = 'red'

          this.leftRotate(x.parent)

          w = x.parent.right
        }

        if (!w.left.isRed() && !w.right.isRed()) {
          w.color = 'red'
          x = x.parent
        }
        //
        else {
          if (!w.right.isRed()) {
            w.left.color = 'black'
            w.color = 'red'

            this.rightRotate(w)

            w = x.parent.right
          }

          w.color = x.parent.color

          x.parent.color = 'black'
          w.right.color = 'black'

          this.leftRotate(x.parent)

          x = this.root
        }
      } else {
        let w = x.parent.left

        if (w.isRed()) {
          w.color = 'black'
          x.parent.color = 'red'

          this.rightRotate(x.parent)

          w = x.parent.left
        }
        if (!w.right.isRed() && !w.left.isRed()) {
          w.color = 'red'
          x = x.parent
        } else {
          if (!w.left.isRed()) {
            w.right.color = 'black'
            w.color = 'red'
            this.leftRotate(w)
            w = x.parent.left
          }

          w.color = x.parent.color

          x.parent.color = 'black'
          w.left.color = 'black'

          this.rightRotate(x.parent)

          x = this.root
        }
      }
    }

    x.color = 'black'
  }

  /**
   *
   * @param z
   */
  insertFixup(z) {
    while (z.parent.isRed()) {
      if (z.parent === z.parent.parent.left) {
        const y = z.parent.parent.right

        if (y.isRed()) {
          z.parent.color = 'black'
          y.color = 'black'
          z.parent.parent.color = 'red'
          z = z.parent.parent
        } else {
          if (z === z.parent.right) {
            z = z.parent
            this.leftRotate(z)
          }

          z.parent.color = 'black'
          z.parent.parent.color = 'red'
          this.rightRotate(z.parent.parent)
        }
      } else {
        const y = z.parent.parent.left

        if (y.isRed()) {
          z.parent.color = 'black'
          y.color = 'black'
          z.parent.parent.color = 'red'
          z = z.parent.parent
        } else {
          if (z === z.parent.left) {
            z = z.parent
            this.rightRotate(z)
          }

          z.parent.color = 'black'
          z.parent.parent.color = 'red'

          this.leftRotate(z.parent.parent)
        }
      }
    }

    this.root.color = 'black'
  }

  /**
   *
   * @param x
   */
  leftRotate(x) {
    const y = x.right

    x.right = y.left

    if (y.left !== this.nil) y.left.parent = x

    y.parent = x.parent

    if (x.parent === this.nil) this.root = y
    else if (x === x.parent.left) x.parent.left = y
    else x.parent.right = y

    y.left = x
    x.parent = y
  }

  /**
   *
   * @param x
   */
  rightRotate(x) {
    const y = x.left
    x.left = y.right

    if (y.right !== this.nil) y.right.parent = x

    y.parent = x.parent

    if (x.parent === this.nil) this.root = y
    else if (x === x.parent.left) x.parent.left = y
    else x.parent.right = y

    y.right = x
    x.parent = y
  }
}

/**
 *
 */
class OrderedSet {
  constructor() {
    this.rbtree = new RBTree()
    this.size = 0
  }

  /**
   *
   * @param key
   * @param value
   */
  insert(key, value) {
    this.rbtree.insert(key, value)
    this.size++
  }

  /**
   *
   * @param key
   * @param value
   */
  find(key, value) {
    let node = this.rbtree.root
    let targetNode = null
    const tempNode = new RBTreeNode(key, value, this.rbtree.nil)

    while (node !== this.rbtree.nil) {
      if (this.rbtree.compare(tempNode, node) === 0) {
        targetNode = node
        break
      }
      //
      else if (this.rbtree.compare(tempNode, node) < 0) node = node.left
      else node = node.right
    }

    return targetNode ? targetNode.value : null
  }

  /**
   *
   * @param key
   * @param value
   */
  delete(key, value) {
    this.rbtree.delete(key, value)
    this.size--
  }

  /**
   *
   */
  getMin() {
    const node = this.rbtree.minimum(this.rbtree.root)

    return [node.key, node.value]
  }

  /**
   *
   */
  getMax() {
    const node = this.rbtree.maximum(this.rbtree.root)

    return [node.key, node.value]
  }

  /**
   *
   */
  inorderTraversal() {
    const result = []

    const inorder = (node) => {
      if (node !== this.rbtree.nil) {
        inorder(node.left)

        result.push([node.key, node.value])

        inorder(node.right)
      }
    }

    inorder(this.rbtree.root)

    return result
  }
}
